### ISSUE TITLE
Node errors are very slow to create

### ISSUE THREAD

**Issue by ronag** [nodejs/performance#40]
Unfortunately, it's becoming increasingly common that errors are used as part of normal control flow.

- `AbortController` throws an `AbortError` as part of cancellation logic.
- A lot of web frameworks use throwing errors as part of request response logic using e.g. https://www.npmjs.com/package/http-errors
- Web specs use throwing errors as a normal flow control mechanism (e.g. WebStreams)

Hence the current state of constructing errors in Node is not optimal. The errors are heavily enriched with lots of helpful information, however at a huge performance cost. I believe we need to investigate what can be done to improve this and discuss possible compromises.

**Comment by ronag** [comment]
@mcollina

**Comment by anonrig** [comment]
I’m aware of this, and it’s on my to-do list. Appreciate any experiences and/or thoughts on this!

**Comment by mcollina** [comment]
This is really problematic. I would recommend we include also @jasnell and maybe we schedule a custom meeting for this. There is a lot to talk about.

**Comment by aduh95** [comment]
FWIW I have [an open PR](https://github.com/nodejs/node/pull/46065) that makes Error creation slower, but also less vulnerable. I guess it'd be nice to see if there's a way to make both things happen or if those are two conflicting goals.

**Comment by BridgeAR** [comment]
I am currently looking into improving the performance. So far I have identified two things that we are able to improve.

We are able to do move some checks from the runtime to the setup. This is however only a small part of the overall performance overhead. One big bottleneck is `ErrorCaptureStackTrace` and this is a tricky part. We might be able to remove that.

Besides the actual performance bottleneck, the implementation itself has become difficult to maintain by now as we have lots of very similar functions that create things slightly different. I am working on unifying that but it is going to take some time to do this right.

**Comment by BridgeAR** [comment]
One note on `AbortError`: these are "simple" errors and they should be on par with other errors.
What we can improve is to remove the stack trace as it should likely not contain any useful data. Creating the stack trace is pretty much always the biggest overhead for errors.

**Comment by mcollina** [comment]
@BridgeAR a quick analysis on my end showed that the bottleneck for generic errors (such as `TypeError`) comes from our override. I don't think we should override the "standard" errors at all.

**Comment by jasnell** [comment]
Yeah the internal/errors.js situation is... Unfortunate. Tons more complexity has evolved there than was originally intended.

**Comment by benjamingr** [comment]
We can't remove stack traces usually (in stuff like AbortError) as it significantly harms debugging. The workaround is to capture them lazily like "regular" errors and move to more standard errors like @mcollina suggests IMO.

**Comment by BridgeAR** [comment]
I already implemented some first improvements locally:

```
                                                                         confidence improvement accuracy (*)   (**)   (***)
error/hidestackframes.js nested=0 n=100000 type='direct-call-noerr'                     -1.66 %       ±5.83% ±7.76% ±10.10%
error/hidestackframes.js nested=0 n=100000 type='direct-call-throw'             ***     35.83 %       ±4.12% ±5.53%  ±7.29%
error/hidestackframes.js nested=0 n=100000 type='hide-stackframes-noerr'                 2.73 %       ±6.35% ±8.45% ±11.00%
error/hidestackframes.js nested=0 n=100000 type='hide-stackframes-throw'        ***     41.35 %       ±3.64% ±4.85%  ±6.31%
error/hidestackframes.js nested=1 n=100000 type='direct-call-noerr'                     -0.59 %       ±4.63% ±6.16%  ±8.02%
error/hidestackframes.js nested=1 n=100000 type='direct-call-throw'             ***     74.44 %       ±2.54% ±3.39%  ±4.42%
error/hidestackframes.js nested=1 n=100000 type='hide-stackframes-noerr'                -0.58 %       ±3.09% ±4.11%  ±5.34%
error/hidestackframes.js nested=1 n=100000 type='hide-stackframes-throw'        ***     73.76 %       ±2.71% ±3.60%  ±4.68%
error/node-error.js type='node' n=100000                                        ***     55.52 %       ±2.20% ±2.93%  ±3.83%
error/node-error.js type='regular' n=100000                                             -0.88 %       ±1.42% ±1.89%  ±2.46%
```

**Comment by BridgeAR** [comment]
I checked AbortError again and it seems like the stack trace is needed in most cases to allow users to identify the point where they called `abort()` or similar.

There are a few places where it's possible to remove the stack frames such as when running in a timeout (`AbortSignal.timeout(ms)`). Besides that, I have only found a case in child process that should actually create the error earlier to include the proper stack trace.

The errors are just slow as long as we have to collect the stack frames and we should probably keep that behavior as it is.

**Comment by H4ad** [comment]
Just to keep things tracked, currently, we have these PRs addressing this issue:

- [ ] https://github.com/nodejs/node/pull/46648
- [ ] https://github.com/nodejs/node/pull/49642
- [x] https://github.com/nodejs/node/pull/49643
- [x] https://github.com/nodejs/node/pull/49654
- [ ] https://github.com/nodejs/node/pull/49696

Except for the first, all of them were made by @Uzlopak.

The first one brings a lot of improvements but hasn't had any activity in months, @BridgeAR do you need any help?

Probably these new changes made by Uzlopak will bring a lot of conflicts, and maybe some of the optimizations will overlap with the changes made by BridgeAR, if we don't receive updates from BridgeAR, I suggest creating new smaller PRs using the first PR as a base just to not lose these big improvements.

**Comment by Uzlopak** [comment]
Created now also 

https://github.com/nodejs/node/pull/49956

as first step to improve SystemError instanciation.

**Comment by Uzlopak** [comment]
I think i found a somewhat elegant way to change hideStackframe so that we dont need to patch the name of the function, and the need to handle those in prepareStackTrace. It should then also make the use of captureLargerStackTrace obsolete. 

On the long run, I try to avoid providing a custom prepareStackTrace, and stack trace output should then be based solely on v8. But i cant promise that. 

One obstacle at a time.

Stay tuned.

**Comment by Uzlopak** [comment]
Sooo.

Created my proposal: https://github.com/nodejs/node/pull/49990

**Comment by Uzlopak** [comment]
Sooo.

Created my proposal: https://github.com/nodejs/node/pull/49990

**Comment by Uzlopak** [comment]
You can set it via a cli flag:
https://nodejs.org/api/cli.html#--stack-trace-limitlimit

If you set it to 0, than it gets really fast
